### ISSUE TITLE
dlib http server error

### ISSUE THREAD

**Issue by kenkit** [davisking/dlib#951]
So I put the example server code in a function other than main. Then I start it via boost thread, but while exiting I get 
```
Error detected at line 38.
Error detected in file e:\programming\dlib-19.7\dlib\threads\threaded_object_extension.cpp.
Error detected in function __thiscall dlib::threaded_object::~threaded_object(void).

Failing expression was is_alive() == false.
        threaded_object::~threaded_object()
        You have let a threaded object destruct itself before terminating its thread
        this: 0110E838

Assertion failed: false, file e:\programming\dlib-19.7\dlib\threads\threaded_object_extension.cpp, line 43
```

**Comment by kenkit** [comment]
I'm building with g++-6
Another issue is I can't run the program in the background or as a script on linux.
I don't know if it's relate to the cin.get being part of the code.
I need a way of starting the server from another thread.

**Comment by davisking** [comment]
Block your thread so it doesn't destruct the server object.  Like call the
server's start() or do some other thing to block and wait for some event
you define.  But just do anything other than let the server object get
destructed.

**Comment by kenkit** [comment]
is this what you mean. I've tried alot of options, this works when main is blocked by cin.get
but when you replicate same code into the main replacing the cin.get assuming something else will set running ==false. I get the same error.
I noticed the crash  on windows.
```
	our_web_server.start();
		while(!running)
		dlib::sleep(100);
```

**Comment by kenkit** [comment]
maybe I should declare our_web_server as a global variable ?

**Comment by kenkit** [comment]
It didn't work

**Comment by kenkit** [comment]
I start it with 

```
boost::thread server(start_server);


	while(!running_threads.get_engine_set())
	dlib::sleep(100); 
	crawlling.join();	
```

**Comment by kenkit** [comment]
Starting the function which contains
```
		our_web_server.start_async();
	while(!running)
	dlib::sleep(100);
```


and 

in main

`start_server();`


stilll causes the crash

**Comment by kenkit** [comment]
Also calling the original example still fails.
I have no other threads within the class.


try
    {
        // create an instance of our web server
        web_server our_web_server;

		// make it listen on port 5000
		#ifdef testing_server
			our_web_server.set_listening_port(5001);
		#else 
			our_web_server.set_listening_port(5002);
		#endif
        // Tell the server to begin accepting connections.
		our_web_server.start_async();
	while(running)
	dlib::sleep(100);

		//our_web_server.clear();
		//cin.get(); 
    }
    catch (exception& e)
    {
        logmain<< LINFO<< e.what() ;
    }

**Comment by kenkit** [comment]
but I do have a logging class with the following code, I have other classes too, but they are not dynamic

```
class Logging
{
public:
    Logging(const std::string& logfile) : fout(logfile.c_str()) { }

    void hook(
        const std::string& logger_name,
        const log_level& l,
        const uint64 thread_id,
        const char* message_to_log
    )
    {
        if (l.priority >= LDEBUG.priority)
            fout      << l.name << " " << logger_name << ": " << message_to_log << std::endl;
        if (l.priority >= LINFO.priority)
            std::cout << l.name << " " << logger_name << ": " << message_to_log << std::endl;
    }

private:
    std::ofstream fout;
};
```

called from main
```
Logging hook("log.txt");;
set_all_logging_output_hooks(hook,&Logging::hook);

  setup_loggers();
```
and 
```
void setup_loggers ()
{


logger logserver("server");
logger logmain("other");


    logserver.set_level(LINFO);
	logmain.set_level(LINFO);

}
```

**Comment by davisking** [comment]
You don't even need a thread.  All you need to do is put the server object
somewhere where it won't be destructed until you are ready to close your
application.  Where you put it is up to you.  You must have other variables
in your program.  It's just another variable, put it with your other
variables.
